pageserver/src/thread_mgr.rs

```diff
@@ -203,9 +220,26 @@ fn thread_wrapper<F, E>(
     // Remove our entry from the global hashmap.
     THREADS.lock().unwrap().remove(&thread_id);
 
-    // If the thread payload panic'd, exit with the panic.
-    if let Err(err) = result {
-        panic::resume_unwind(err);
+    match result {
+        Ok(Ok(())) => info!("Thread '{}' exited normally", thread_name),
+        Ok(Err(err)) => {
+            if fail_on_error {
+                error!(
+                    "Shutting down: thread '{}' exited with error: {:?}",
+                    thread_name, err
+                );
+                shutdown_pageserver();
```

Comment:
I wonder, do we really need to kill whole pageserver if e g gc for one of the tenants failed? It could help in resolving the error, but neighbours might be a bit unhappy with that. Maybe we need two flags, shutdown only affected tenant, or whole pageserver. Whole pageserver should be restarted if pg or http protocol listeners are dead for example. Another more sophisticated option is to have some number of retries, and when thread fails we try to shutdown tenant first, and if it continues to fail then try to kill whole pageserver. WDYT? Maybe it is too complex and we do not need that :)

Reply:
> Whole pageserver should be restarted if pg or http protocol listeners are dead for example

I've implemented the logic to continue working in those two cases in particular, by the way. Should I do otherwise?
Feels like we could also restart those particular threads?

> do we really need to kill whole pageserver if e g gc for one of the tenants failed

I do agree that we might have something more sophisticated, presumably with restarts, but that would need better design, IMO.
Main point is that this has to be done outside of `thread_mgr`, since every thread will have its own notion on how to recover, restart, etc.

In this PR, I think the log explicitness and the correctness is the main point: if we determine which threads' errors is ok to skip and what to do with the others' issues, we can increment from that later.